translation works by LiuNing (#364)
Oneflow-Inc/oneflow-documentation: en/docs/basics/06_optimization.md

> # Back propagation and optimizer

Author: 已修改

> So far, we have learned how to use OneFlow to [load data](./03_dataset_dataloader.md),[build models](./04_build_network.md),[calculate gradient of model parameters](./05_autograd.md), and combine them so that we can train models by using backpropagation algorithms.

Author: 均已修改

> In [oneflow.optim](https://oneflow.readthedocs.io/en/master/optim.html) ,there are various `optimizer`that simplify the code of back propagation.

Author: 已修改

This article will first introduce the basic concepts of back propagation and then show you how to use the `oneflow.optim` class.

> ## numpy 手工实现反向传播

Author: 已添加

> In order to make it easier for readers to understand the relationship between back propagation and automatic derivation, a training process of a simple model implemented only with numpy is provided here:

Reviewer: ```suggestion
In order to make it easier for readers to understand the relationship between backpropagation and autograd, a training process of a simple model implemented with numpy is provided here:
```

backpropagation 用作一个词更好？我看 PyTorch Tutorial 是这样做的。用 autograd 更好是因为这是我们系列教程中出现过的内容。这也是为什么说 **先通读系列教程**

only 去掉是因为不必要直译，翻译完成后请主动认真阅读自己的翻译效果，做调整，达到自己认为最佳的效果。

Author: 已修改和基本阅读了cn文档

import numpy as np

ITER_COUNT = 500
LR = 0.01

# forward propagation
def forward(x, w):
    return np.matmul(x, w)

# loss function
def loss(y_pred, y):
    return ((y_pred - y) ** 2).sum()

# calculate gradient
def gradient(x, y, y_pred):
    return np.matmul(x.T, 2 * (y_pred - y))

if __name__ == "__main__":
    # train: Y = 2*X1 + 3*X2
    x = np.array([[1, 2], [2, 3], [4, 6], [3, 1]], dtype=np.float32)
    y = np.array([[8], [13], [26], [9]], dtype=np.float32)

    w = np.array([[2], [1]], dtype=np.float32)
    # training cycle
    for i in range(0, ITER_COUNT):
        y_pred = forward(x, w)
        l = loss(y_pred, y)
        if (i + 1) % 50 == 0:
            print(f"{i+1}/{500} loss:{l}")

        grad = gradient(x, y, y_pred)
        w -= LR * grad

    print(f"w:{w}")
```

output：

```text
50/500 loss:0.0034512376878410578
100/500 loss:1.965487399502308e-06
150/500 loss:1.05524122773204e-09
200/500 loss:3.865352482534945e-12
250/500 loss:3.865352482534945e-12
300/500 loss:3.865352482534945e-12
350/500 loss:3.865352482534945e-12
400/500 loss:3.865352482534945e-12
450/500 loss:3.865352482534945e-12
500/500 loss:3.865352482534945e-12
w:[[2.000001 ]
 [2.9999993]]
```

> Note that the loss function expression we selected is  $\sum (y_{p} - y)^2$,so the code for gradient of `loss` to parameter `w` is：

Reviewer: ```suggestion
Note that the loss function expression we selected is $\sum (y_{p} - y)^2$, so the code for gradient of `loss` to parameter `w` is：
```

```python
def gradient(x, y, y_pred):
    return np.matmul(x.T, 2 * (y_pred - y))
```

[SGD](https://en.wikipedia.org/wiki/Stochastic_gradient_descent) is used to update parameters：

```python
grad = gradient(x, y, y_pred)
w -= LR*grad
```

In summary, a complete iteration in the training includes the following steps:

1. The model calculates the predicted value based on the input and parameters (`y_pred`)

> 2. Calculate loss,which is the error between the predicted value and the label

Reviewer: ```suggestion
2. Calculate loss, which is the error between the predicted value and the label
```

> 3. calculate the gradient of loss to parameter

Reviewer: ```suggestion
3. Calculate the gradient of loss to parameter(s)
```

> 4. Update parameter

Reviewer: ```suggestion
4. Update parameter(s)
```

1 and 2 are forward propagation process; 3 and 4 are back propagation process.

## Hyperparameters

> Hyperparameters are parameters related to model training settings, which can affect the efficiency and results of model training.As in the above code `ITER_COUNT`,`LR` are super parameters.

Reviewer: ```suggestion
Hyperparameters are parameters related to model training settings, which can affect the efficiency and results of model training.As in the above code `ITER_COUNT`, `LR` are 【超参】.
```

为什么会把超参翻译诚 super parameters？

## Using the optimizer class in `oneflow.optim`

> Using the optimizer class in `oneflow.optim` for back propagation will be more concise. Next, we show how to use it.

Reviewer: ```suggestion
Using the optimizer class in `oneflow.optim` for back propagation will be more concise.
```
不一定要直译，比如翻译成英文时，感觉后面的话多余了，就可以去掉。

> First, prepare the data and model. The convenience of using Module is that you can place the super parameters in Module for management.

Reviewer: 超参

Author: 已修改

```python
import oneflow as flow

x = flow.tensor([[1, 2], [2, 3], [4, 6], [3, 1]], dtype=flow.float32)
y = flow.tensor([[8], [13], [26], [9]], dtype=flow.float32)

class MyLrModule(flow.nn.Module):
    def __init__(self, lr, iter_count):
        super().__init__()
        self.w = flow.nn.Parameter(flow.tensor([[2], [1]], dtype=flow.float32))
        self.lr = lr
        self.iter_count = iter_count

    def forward(self, x):
        return flow.matmul(x, self.w)

model = MyLrModule(0.01, 500)
```

> ### loss function

Reviewer: 大小写

Author: 已修改

Then, select the loss function. OneFlow comes with a variety of loss functions. We choose [MSELoss](https://oneflow.readthedocs.io/en/master/nn.html?highlight=mseloss#oneflow.nn.MSELoss) here：

```python
loss = flow.nn.MSELoss(reduction="sum")
```
### Construct optimizer

> The logic of back propagation is encapsulated in optimizer.We choose [SGD](https://oneflow.readthedocs.io/en/master/optim.html?highlight=sgd#oneflow.optim.SGD) here, You can choose other optimization algorithms as needed, such as [Adam](https://oneflow.readthedocs.io/en/master/optim.html?highlight=adam#oneflow.optim.Adam) and[AdamW](https://oneflow.readthedocs.io/en/master/optim.html?highlight=adamw#oneflow.optim.AdamW) .

Reviewer: 空格。

The logic of back propagation is encapsulated in optimizer 这句可能不够自然，需要 @wwd190906 @BONNNIE review 时重点确认下是否要改，怎么改

```python
optimizer = flow.optim.SGD(model.parameters(), model.lr)
```

> When constructing `optimizer`, pass model parameters and learning rate to `SGD`. Then call `optimizer.step()`, and it automatically completes the gradient of the model parameters and updates the model parameters according to the SGD algorithm.

Reviewer: 中文没有主语是自然的。英文大量没有主语会很不自然。

When the `optimizer` is constructed, the model pramters and learning rate is given. Then the `optimizer.step()` is called.

翻译的好坏问题没有特别死板的标准，但是又是客观存在的，需要自己多参考资料，琢磨

Author: 已修改

### Train

> When the above preparations are completed, you can start training:

Reviewer: 全文的视角要统一，不要没有特殊理由的情况洗，随意切换 "we" 和 "you"
